src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newStates = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        const StateBfAdd = { ...(newStates[newStates.length - 1] || state) };
+
+        newStates.push({ ...StateBfAdd, ...action.extraData });
+        break;
+      case 'removeProperties':
+        const stateBfRemove = { ...(newStates[newStates.length - 1] || state) };
+
+        action.keysToRemove.forEach((key) => {
+          if (Object.keys(stateBfRemove).includes(key)) {
+            delete stateBfRemove[key];
+          }
+        });
+        newStates.push({ ...stateBfRemove });
+        break;
+      case 'clear':
+        newStates.push({});
```

Comment:
[CODE STYLE] - Directly pushing the new state into 'newStates' array is fine, but to improve readability and maintainability, consider assigning the new state to a variable first and then push it to the array.